certificate/certificates.go

```diff
@@ -287,24 +288,51 @@ func (c *Certifier) getForCSR(domains []string, order acme.ExtendedOrder, bundle
 // The certRes input should already have the Domain (common name) field populated.
 //
 // If bundle is true, the certificate will be bundled with the issuer's cert.
-func (c *Certifier) checkResponse(order acme.Order, certRes *Resource, bundle bool) (bool, error) {
+func (c *Certifier) checkResponse(order acme.Order, certRes *Resource, bundle bool, preferredChain string) (bool, error) {
 	valid, err := checkOrderStatus(order)
 	if err != nil || !valid {
 		return valid, err
 	}
 
-	cert, issuer, err := c.core.Certificates.Get(order.Certificate, bundle)
-	if err != nil {
-		return false, err
+	var (
+		links      = append([]string{order.Certificate}, order.AlternateChainLinks...)
+		defaultRes *Resource
+	)
+	for _, link := range links {
+		cert, issuer, err := c.core.Certificates.Get(link, bundle)
+		if err != nil {
+			return false, err
+		}
+		if defaultRes == nil {
+			defaultRes = &Resource{
+				IssuerCertificate: issuer,
+				Certificate:       cert,
+				CertURL:           link,
+				CertStableURL:     link,
+			}
+		}
+		if preferredChain != "" {
+			x509Certs, err := certcrypto.ParsePEMBundle(issuer)
+			if err != nil {
+				return false, err
+			}
+			if x509Certs[0].Subject.CommonName != preferredChain {
+				continue
```

Comment:
As written, a user would not be able to do this:

    --preferred-chain="DST Root CA X3"

because the root certificate is not included in the chain, and as such, no `.Subject` would ever match.

This could be addressed by instead matching against the `.Issuer` field of *every* certificate in the bundle. This is how Certbot does it, though other clients have opted for even more ways to match (like SKI/AKI/hash).

Interactive example demonstrating this use-case against this PR:

    $ go run cmd/lego/main.go --server https://acme-staging-v02.api.letsencrypt.org/directory -d 976fb1d7fc17.ngrok.io --http --http.port=":8123" -m red@act.ed run --preferred-chain "Fake LE Root X2"
    2020/08/18 12:38:40 [INFO] [976fb1d7fc17.ngrok.io] acme: Obtaining bundled SAN certificate
    2020/08/18 12:38:41 [INFO] [976fb1d7fc17.ngrok.io] AuthURL: https://acme-staging-v02.api.letsencrypt.org/acme/authz-v3/95951549
    2020/08/18 12:38:41 [INFO] [976fb1d7fc17.ngrok.io] acme: authorization already valid; skipping challenge
    2020/08/18 12:38:41 [INFO] [976fb1d7fc17.ngrok.io] acme: Validations succeeded; requesting certificates
    2020/08/18 12:38:42 [INFO] lego has been configured to prefer certificate chains with issuer "Fake LE Root X2", but no chain from the CA matched this issuer. Using the default certificate chain instead.